### PR TITLE
Add release workflow for che parent and server

### PR DESCRIPTION
**Pull request by mkuznyetsov**
### What does this PR do?
Move che parent & server release workflow from che-release project to che-server 

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18780

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      - uses: actions/checkout@v2
+        with: 
+          fetch-depth: 0
+          path: che
+      - uses: actions/checkout@v2
+        if: ${{ github.event.inputs.releaseParent == 'true' }}
+        with:
+          repository: eclipse/che-parent
+          token: {{ secrets.CHE_BOT_GITHUB_TOKEN }}
+          path: che-parent
```
Comment (nickboldt) [inline .github/workflows/release.yml:53]: does this need `fetch-depth:0` too?

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      - name: Check existing tag
+        run: |
+          if [[ "$FORCE_RECREATE_TAGS" == "false" ]] && [[ $(cd che && git ls-remote --exit-code origin refs/tags/${{ github.event.inputs.version}} ]]; then
+            echo "cannot create release, when tag already exists"
```
Comment (nickboldt) [inline .github/workflows/release.yml:58]: Might be better formatted as an error, eg.,

```
echo "[ERROR] Tag ${{ github.event.inputs.version}} already exists - cannot re-release. Use 'forceRecreateTags:true' if you want to force a re-release to this tag."

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        with:
+          python-version: 3.6
+      - uses: actions/setup-java@v1
+        with:
+          java-version: '11.0.7'
+          java-package: jdk
+          architecture: x64 
+      - name: Set up environment
+        run: |
+          # install more repos
```
Comment (nickboldt) [inline .github/workflows/release.yml]: update doesn't install repos... it updates.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+          sudo apt-get update -y || true
+          # install more dependencies
+          sudo apt-get -y -q install wget curl bash git 
+          java -version
```
Comment (nickboldt) [inline .github/workflows/release.yml:87]: you're checking java version here but didn't install it?

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+          # want >=5
+          bash --version
+
+          java -version
+          which java
+          javac -version
+          which javac
+
+          # do we need any of these?
```
Comment (nickboldt) [inline .github/workflows/release.yml]: if these were commented out in the last che 7.25.x releases, we can probably remove them, right?

File: make-release.sh [hunk 6]
```diff
@@ -3,162 +3,405 @@
 # Used to create branch/tag, update versions in pom.xml
 # and and trigger release by force pushing changes to the release branch
```
Comment (nickboldt) [inline make-release.sh]: are we still using the release branch?

File: make-release.sh [hunk 7]
```diff
@@ -3,162 +3,405 @@
...
-[[ $VERSION =~ ^([0-9]+)\.([0-9]+)\.([0-9]+) ]] && BASE="${BASH_REMATCH[1]}.${BASH_REMATCH[2]}"; NEXT="${BASH_REMATCH[3]}"; (( NEXT=NEXT+1 )) # for VERSION=7.7.1, get BASE=7.7, NEXT=2
-NEXTVERSION_Z="${BASE}.${NEXT}-SNAPSHOT"
-bump_version ${NEXTVERSION_Z} ${BUGFIX}
-
-popd > /dev/null || exit
-
-# cleanup tmp dir
-# cd /tmp && rm -fr "$TMP"
+
+# TODO ensure usage of respective bugfix branches
```
Comment (nickboldt) [inline make-release.sh]: what does this mean?

File: make-release.sh [hunk 8]
```diff
@@ -3,162 +3,405 @@
...
+    gpg --import $HOME/.m2/gpg.key
+    # gpg --import --batch $HOME/.m2/gpg.key
+    gpg --version
+}
+
+installDebDeps(){
+    set +x
+    # TODO should this be node 12?
+    curl -sL https://deb.nodesource.com/setup_10.x | sudo -E bash -
+    sudo apt-get install -y nodejs
```
Comment (nickboldt) [inline make-release.sh:54]: should this move to a yaml step?

**Review by nickboldt — APPROVED** [review]
some inline questions but probably OK to merge.

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,116 @@
+#
+# Copyright (c) 2020 Red Hat, Inc.
```
Comment (nickboldt) [inline .github/workflows/release.yml:2]: omg 2020 was SOOO last year. :)

File: make-release.sh (hunk 6 above)
Comment (mkuznyetsov) [reply make-release.sh]: fixed

File: .github/workflows/release.yml (hunk 1 above)
Comment (mkuznyetsov) [reply .github/workflows/release.yml:53]: fixed

File: .github/workflows/release.yml (hunk 4 above)
Comment (nickboldt) [reply .github/workflows/release.yml:87]: sure but you're ALSO outputting version 3 lines lower.

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -31,7 +31,6 @@ jobs:
     steps:
       - uses: actions/checkout@v2
         with: 
-          fetch-depth: 0
```
Comment (nickboldt) [inline .github/workflows/release.yml]: I was suggesting that we NEED `fetch-depth: 0` for both checkouts, not to remove it from the first one. Without it, you get only the last sha commit and can't switch to any branch as none exist. Am I wrong?

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -17,13 +17,25 @@ on:
...
+      versionParent:
+        description: 'Specify che parent version here, if its release is enabled'
+        required: false
+        default: ''
       forceRecreateTags:
         description: 'If true, tags will be overriden and regenerated. Use with caution'
         required: false
         default: 'false'
+      autoReleaseOnNexus:
+        description: 'If true, artifacts will be automatically released when deployed to Nexus'
```
Comment (nickboldt) [inline .github/workflows/release.yml]: `... when deployed to Nexus, but only if *deployOnNexus* is also true.`

File: .github/workflows/release.yml (hunk 10 above)
Comment (mkuznyetsov) [reply .github/workflows/release.yml]: you're right, it must be added for both

**Pull request merged**